sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCheckResult.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+/**
+ * Represents the result of `Expression.checkInputDataTypes`.
+ * We will throw `AnalysisException` in `CheckAnalysis` if error message is not null.
+ * Use [[TypeCheckResult.success]] and [[TypeCheckResult.fail]] to instantiate this.
+ *
+ */
+class TypeCheckResult private (val errorMessage: String) extends AnyVal {
+  def hasError: Boolean = errorMessage != null
+}
+
+object TypeCheckResult {
+  val success: TypeCheckResult = new TypeCheckResult(null)
+  def fail(msg: String): TypeCheckResult = new TypeCheckResult(msg)
+}
```

Comment:
I was thinking about using value class to save some allocation, but that's probably premature optimization. I think it'd make sense to follow your suggestion (maybe remove the Object, and just name it TypeCheckSuccess and TypeCheckFailure)